### PR TITLE
Fix #9829: Allow `as` in place of `@` for pattern bindings

### LINKED ISSUE #9829: Proposal: Change `@` to `as`
I have a late syntax change proposal: Change `@` as the pattern matching binder to `as`. 

Example:

```scala
  x match 
    case s as Success(x, in1) => ...
    case f as Failure(_, in1) => ...

  e match
    case (xs as hd :: tl, y) => ...
```

Why `@` is not a good choice for pattern matching:

 - It is an obscure symbol with a different meaning. I find the two worst operators to use in programming languages are `@` and `$` since both have everyday uses that have nothing to do with their role in programming. We have avoided `$` so far, let's get rid of `@` as well. 

 - It comes with the wrong precedence. `@` binds more weakly than every other infix operator, yet being a non-standard symbol it would suggest that it binds more strongly.

By contrast, `as` is readable and it comes with the right precedence. It also reinforces learning in relation to givens. In both cases `x as` would be an optional label. 

An objection could be that in patterns, `as`  would be followed by a pattern whereas in givens it's followed by a type. But there's precedence for that: `|` plays a similar dual role: it connects patterns in patterns and types in types.

**Migration** 

For 3.0, allow both `as` and `@`. For 3.1. either deprecate `@` or disallow it altogether and offer an automatic rewrite.

### PR DESCRIPTION
**Pull request by odersky**
For 3.0: Can use either `@` or `as`. 
For 3.1: `@` is deprecated.

### CODE REVIEW

**Comment by som-snytt** [comment]
"At-less shrugged."

**Comment by AdamPaynter** [comment]
I rather like this change. `as` definitely carries more explanatory value than `@`. My only tiny stumbling block is _where_ the `as` keyword appears. Wouldn't it make more sense to say `Pattern(...) as myVar` instead of `myVar as Pattern(...)`? Don't get me wrong, if I could happily accept an arbitrary `@`, I can more than happily accept the `as` presented here. I was just curious if you could help me understand the choice.

**Comment by som-snytt** [comment]
@AdamPaynter the issue suggests https://dotty.epfl.ch/docs/reference/contextual/givens.html as a model for the order. I had the same reservation, given usages in other domains. But if I say, "I took you for a bounder," then it's the alias taken as a bounder, the underlying type. Similarly for any definition where the lhs introduces the binding and the rhs the definition.

**Comment by odersky** [comment]
I think it can go both ways. ML uses `<pattern> as <var>`. By contrast Visual Basic uses `<var> As <type>` as a binder (where we would use `:`).

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 1]
```diff
@@ -2631,7 +2631,10 @@ object Parsers {
     /**  Pattern2    ::=  [id `@'] InfixPattern
      */
     val pattern2: () => Tree = () => infixPattern() match {
-      case p @ Ident(name) if in.token == AT =>
+      case p @ Ident(name) if in.token == AT || in.isIdent(nme.as) =>
+        if in.token == AT && sourceVersion.isAtLeast(`3.1`) then
+          deprecationWarning(s"`@` bindings have been deprecated; use `as` instead", in.offset)
+
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala:2637]: We need to update the `syntax.md` file with the syntax change.

**Review by liufengyun — APPROVED** [review]
LGTM

**Comment by som-snytt** [comment]
Worth observing that this change encapsulates in miniature all that is laudable in the Dotty effort: a judicious willingness to continue to evolve, when others might have rested.

**Comment by nafg** [comment]
Perhaps, if you believe in punctuated equilibrium

**Pull request merged**